coordinator/ethereum/gas-pricing/static-cap/src/main/kotlin/net/consensys/linea/ethereum/gaspricing/staticcap/HistoricVariableCostProviderImpl.kt

```diff
@@ -0,0 +1,74 @@
+package net.consensys.linea.ethereum.gaspricing.staticcap
+
+import linea.OneKWei
+import linea.domain.BlockParameter
+import linea.domain.BlockParameter.Companion.toBlockParameter
+import linea.kotlin.encodeHex
+import linea.web3j.ExtendedWeb3J
+import net.consensys.linea.ethereum.gaspricing.HistoricVariableCostProvider
+import net.consensys.linea.ethereum.gaspricing.MinerExtraDataV1
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+import tech.pegasys.teku.infrastructure.async.SafeFuture
+import java.util.concurrent.atomic.AtomicReference
+
+class HistoricVariableCostProviderImpl(
+  private val web3jClient: ExtendedWeb3J,
+) : HistoricVariableCostProvider {
+  private val log: Logger = LogManager.getLogger(this::class.java)
+  private var lastVariableCost: AtomicReference<Pair<ULong, Double>> =
+    AtomicReference(0UL to 0.0)
+
+  private fun getHistoricVariableCostInWei(blockParameter: BlockParameter): SafeFuture<Double> {
+    return web3jClient.ethGetBlock(blockParameter)
+      .thenApply { block ->
+        if (block == null) {
+          log.warn(
+            "Block {} not found, the requesting node could be out-of-sync",
+            blockParameter.getNumber(),
+          )
+          throw IllegalStateException("Block ${blockParameter.getNumber()} not found")
```

Comment:
nit: although it's supposed to be called with a number, `blockParameter.getNumber()` may throw if it's a TAG.   So just delegate to the underlying to String and it will solve it for us :) 
```suggestion
          throw IllegalStateException("Block $blockParameter not found")
```